From 24ddad86e6e44e5018e3760d31a400b54d459193 Mon Sep 17 00:00:00 2001
From: Antonios Vamporakis <ant@area128.com>
Date: Fri, 3 Jan 2014 19:00:52 +0100
Subject: MIPS: lantiq: arx100_switch: add RMII and RMII_OC phy mode support

Added support for RMII and RMII_OC (output clock) phy modes.

Signed-off-by: Antonios Vamporakis <ant@area128.com>

diff --git a/arch/mips/include/asm/arch-arx100/switch.h b/arch/mips/include/asm/arch-arx100/switch.h
index 301056c..a57a084 100644
--- a/arch/mips/include/asm/arch-arx100/switch.h
+++ b/arch/mips/include/asm/arch-arx100/switch.h
@@ -49,8 +49,10 @@ static inline void build_check_ar9_registers(void)
 #define SW_GCTL0_SE		(1 << 31)
 
 #define RGMII_CTL_P1_SHIFT	10
-#define RGMII_CTL_P1_MASK	(0x3FF << RGMII_CTL_P1_SHIFT)
-#define RGMII_CTL_P0_MASK	0x3FF
+#define RGMII_CTL_P1_MASK	((0x3FF << RGMII_CTL_P1_SHIFT) | RGMII_CTL_P1_OC)
+#define RGMII_CTL_P0_MASK	(0x3FF | RGMII_CTL_P0_OC)
+#define RGMII_CTL_P1_OC		(1 << 23)
+#define RGMII_CTL_P0_OC		(1 << 21)
 #define RGMII_CTL_P0IS_SHIFT	8
 #define RGMII_CTL_P0IS_RGMII	(0x0 << RGMII_CTL_P0IS_SHIFT)
 #define RGMII_CTL_P0IS_MII	(0x1 << RGMII_CTL_P0IS_SHIFT)
diff --git a/drivers/net/lantiq_arx100_switch.c b/drivers/net/lantiq_arx100_switch.c
index cc65249..8048b93 100644
--- a/drivers/net/lantiq_arx100_switch.c
+++ b/drivers/net/lantiq_arx100_switch.c
@@ -3,7 +3,6 @@
  *
  * SPDX-License-Identifier:	GPL-2.0+
  */
-#define DEBUG
 #include <common.h>
 #include <malloc.h>
 #include <netdev.h>
@@ -301,6 +300,31 @@ static void ltq_eth_port_config(struct ltq_eth_priv *priv,
 		}
 
 		break;
+	case PHY_INTERFACE_MODE_RMII:
+		port_xmii = RGMII_CTL_P0IS_RMII;
+		
+		if (!(port->flags & LTQ_ETH_PORT_PHY)) {
+			port_xmii |= (RGMII_CTL_P0SPD_100 |
+			RGMII_CTL_P0DUP_FULL);
+			port_ctl |= P0_CTL_FLP;
+		}
+		
+		break;
+	case PHY_INTERFACE_MODE_RMII_OC:
+		port_xmii = RGMII_CTL_P0IS_RMII;
+		
+		if (!(port->flags & LTQ_ETH_PORT_PHY)) {
+			port_xmii |= (RGMII_CTL_P0SPD_100 |
+			RGMII_CTL_P0DUP_FULL);
+			port_ctl |= P0_CTL_FLP;
+		}
+		
+		if (port->num == 1)
+			port_xmii |= RGMII_CTL_P1_OC;
+		else
+			port_xmii |= RGMII_CTL_P0_OC;
+		
+		break;
 	default:
 		break;
 	}
-- 
2.7.4

